### PR TITLE
feat!: replace `mlh-tsd` with `tsd-lite`

### PR DESCRIPTION
**Pull request by mrazauskas**
Should unblock facebook/jest#11142 and facebook/jest#12198. Also this change would allow to uncomment few assertions of facebook/jest#11949. This is a lot! That is why I though it is time for `tsd-lite`.

The type testing logic is the same between `tsd`, `mlh-tsd` and `tsd-lite`.

The aim of `tsd-lite` was to remove logic unrelated with type testing. For instance, `tsd` and `mlh-tsd` are making few more checks like: if `.d.ts` file exist, if it is included in `package.json` `files` property, etc. These checks and rules are removed in `tsd-lite`.

Additionally `tsd-lite` is returning `fileText` for each failed assertion. TS compiler includes file contents with each diagnostic, so runner does not have to read files again.

Another feature I like – `tsd-lite` is using nearest `tsconfig.json` to read compiler options for each test file. This is close to how `tsc` works. User can have configuration for the whole project, or a group of test file, or just one test file. With `"extends"` option that’s incredibly flexible. For comparison, vanilla `tsd` is only using `tsconfig.json` from the root of the project, it also allows to add TS compiler options to `package.json` and has some defaults which cannot be overridden. Hm.. Might be good for a library written in JS, but sounds complicated and unnecessary for a TS project.

All differences are documented in `tsd-lite` repo – https://github.com/mrazauskas/tsd-lite

### Test plan

I added test to cover both JS and TS projects.

### Todo

Last but not least, I have to send another PR to update Readme file.

### CODE REVIEW

File: package.json
Comment (mrazauskas) [inline package.json:42]: Looked like a good idea to allow user to choose any version.

**Comment by SimenB** [comment]
all of these features seems like things upstream `tsd` should support (turning off/skipping validation) or even straight up features (looking at closest tsconfig, which is the correct thing to do).

That said, happy to land them here now! Would you consider this a breaking change? I guess so due to the peer dep requirement (which is a good idea 👍 )

**Review by SimenB — APPROVED** [review]
this is great, thanks!

Could you update the readme as well? At least mention the peer dependency

**Comment by mrazauskas** [comment]
> all of these features seems like things upstream `tsd` should support (turning off/skipping validation) or even straight up features (looking at closest tsconfig, which is the correct thing to do).

Indeed, can’t agree more.

The changes breaking. For instance, `"stick": true` must be set explicitly with `tsd-lite`. The vanilla `tsd` is the opposite – `"stick": false` must be set explicitly to disable strict checks. In a way, `tsd-lite` is switching defaults. That must be major release.

I will add a note on strictness to `readme`. Working on it.

**Comment by mrazauskas** [comment]
Just fixed documentation. That’s obviously breaking change.

Also `@type` pragma is deprecated. It did not make sense to mention it in the `readme`, but perhaps it is worth to add a line to release notes.

File: README.md
Comment (SimenB) [inline README.md:16]: does this mean if you have an `index.d.ts` file in root it cannot have its own config?

Comment (mrazauskas) [reply README.md:16]: In this case TS compiler configure is used for tests. These will live in separate directory most probably. So the root `tsconfig.json` may apply or can be extend or even overridden by adding a config to that directory. I have to rewrite this paragraph. Thanks for the note.

Comment (SimenB) [reply README.md:16]: right, just wondering if we should accept a path somehow so people can point to a separate config file if they want... probably not needed

Comment (mrazauskas) [reply README.md:16]: Isolate a test file in a directory, drop a specific `tsconfig.json` there. Seems simple and clear. Later just add one more test file to that folder and all works.

Could be there are cases when accepting a path to config file is necessary, but I would wait to see if someone will open an issue.

It could be good idea to have `Configuration` section in the readme with more details. I will think about it.

File: README.md
Comment (mrazauskas) [inline README.md:45]: This is also unclear. An additional section on `Configuration` would do better job.

**Comment by mrazauskas** [comment]
Some more improvement. The vanilla `tsd` does not report if an error in `tsconfig.json` is encountered. Also any syntax errors found in the test files are returned together with test results. Hm.. I think in both of these case it is better to fail early. Otherwise the test result may be unpredictable.

`tsd-lite` now returns `tsdErrors` (alongside with `tsdResults`). So that the runner could throw `Test suite failed to run` if some syntax errors or issues with `tsconfig.json` are encountered.

Could not finish updating documentation without this improvement. Will move on with docs now.

File: package.json
Comment (mrazauskas) [inline package.json:41]: Yarn v2 cannot install `typescript` 4.5.4, but looks like it can install `@tsd/typescript` 4.5.4. Hm.. I think that’s fine. Or?

File: src/run.js
Comment (mrazauskas) [inline src/run.js:18]: Strange line. I was looking at Jest code, this was enough to print nice message with `Test suite failed to run` title on top. Alternatively it is possible to throw an error in `tsd-lite`. The result would be almost the same: if errors are formatted here, it is possible to report several errors at once (see tests); if `tsd-lite` would throw, only one error could be reported.

Comment (mrazauskas) [reply src/run.js:18]: I will rework this. Tried it out, all shapes up much better if `tsd-lite` throws an error.

**Comment by SimenB** [comment]
@mrazauskas sorry, lost track of this. Do you think it's ready? 😀

**Comment by mrazauskas** [comment]
In general it is ready. I have one more small change in `tsd-lite`. Will publish new version today and will bump it here.

At the same time, I was in doubt. The readme still sounds like this is a workaround. So I was fine to leave it to cool down a bit. And to ask again: does it look good? Hm.. Perhaps this is alright as a temporary solution?

Still working on what is mentioned in #32. Had to rewrite my parser completely, now `only`, `skip` and `todo` work as expected with `describe` and `test`. Also `-t` flag and watch mode (with `run only failed`). And much more.

I was trying to migrate some parts of Jest’s type tests to my new testing tool. Looked very promising. There are cases which do not work yet. Polishing it. Should be finished in few months. So this PR is somewhat temporary solution. At least that’s how I see it. Could be merged, but it can wait too.

Might be good idea to unblock the issues which I was mentioning in the first comment. How do you feel about it? (;

**Comment by SimenB** [comment]
I'm hoping to have a burst of energy soonish to land a new major of Jest, at which point I wanna land an update of TS. So I think landing this now, while waiting for #32 makes sense - it's still an improvement over what we have today 🙂

File: README.md
Comment (SimenB) [inline README.md:22]:
```suggestion
Remember to install `@tsd/typescript` package matching your version of `typescript`. It is a required peer dependency.
```

File: README.md
Comment (SimenB) [inline README.md:1]: why?

File: README.md
Comment (SimenB) [inline README.md:19]: I think keeping both `--dev` and the npm script makes sense

File: README.md
Comment (SimenB) [inline README.md:28]:
```suggestion
First of all, you should [configure Jest](https://jestjs.io/docs/configuration) to discover your test files. For example, if the files has the suffix `.test.ts` and live inside `__typetests__` directory, set up `jest.config.tsd.js` like this:
```

File: README.md
Comment (SimenB) [inline README.md:43]:
```suggestion
Your test files will be compiled using the [TypeScript compiler](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API) (similar to `tsc`), but, instead of emitting JS code, `tsd-lite` will analyze types and diagnostics returned by the compiler.
```

**Review by SimenB — APPROVED** [review]
LGTM, just some nits for the readme 🙂

File: README.md
Comment (mrazauskas) [reply README.md]: Set in monospace font it was too long for mobile screen ;D Seemed like bold is enough. Not important at all. Can stay.

**Comment by mrazauskas** [comment]
@SimenB All should be fixed. Take a look, please.

Glad to hear you are think about a new major of Jest. Node 10 support also becomes more and more blocking. Time to move on?

**Comment by SimenB** [comment]
Yep, dropping node 10 and 15 🙂

File: README.md
Comment (SimenB) [inline README.md:19]:
```suggestion
yarn add --dev jest-runner-tsd @tsd/typescript
```

**Review by SimenB — APPROVED** [review]
thanks!

**Comment by SimenB** [comment]
https://github.com/jest-community/jest-runner-tsd/releases/tag/v2.0.0

**Comment by SimenB** [comment]
Can do a quick v3 dropping node 10?

**Comment by mrazauskas** [comment]
Sure. Yes, I can do. Also Renovate will be happier.

**Comment by SimenB** [comment]
Would you mind sending a PR to jest updating to v2?

**Comment by mrazauskas** [comment]
> Would you mind sending a PR to jest updating to v2? then we'll do v3 when I land [facebook/jest#12220](https://github.com/facebook/jest/pull/12220)

Ok. Will do this.

**Pull request merged**
